diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt

```diff
@@ -491,35 +494,98 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
-            val isNewLinePrev = it.treePrev?.isNewLineNode() ?: false
-            (it.elementType == COMMA && !isNewLineNext) ||
-                    // Move RPAR to the new line
-                    (it.elementType == RPAR && it.treePrev?.elementType != COMMA && !isNewLinePrev)
+    private fun isCorrectSuperTypeList(valueParameterList: List<ASTNode>): Boolean {
+        val superTypeList = valueParameterList.filter { it.elementType in listOf(SUPER_TYPE_CALL_ENTRY, SUPER_TYPE_ENTRY) }
+        if (superTypeList.size <= 2) {
+            return true
         }
-        .toList()
-        .takeIf { it.isNotEmpty() }
-        ?.let { invalidCommas ->
-            val warnText = node.getParentIdentifier()?.let {
-                "$entryType should be placed on different lines in declaration of <${node.getParentIdentifier()}>"
-            } ?: "$entryType should be placed on different lines"
-            WRONG_NEWLINES.warnAndFix(configRules, emitWarn, isFixMode,
-                warnText, node.startOffset, node) {
-                invalidCommas.forEach { commaOrRpar ->
-                    val nextWhiteSpace = commaOrRpar.treeNext?.takeIf { it.elementType == WHITE_SPACE }
-                    if (commaOrRpar.elementType == COMMA) {
-                        nextWhiteSpace?.treeNext?.let {
-                            commaOrRpar.appendNewlineMergingWhiteSpace(nextWhiteSpace, nextWhiteSpace.treeNext)
-                        } ?: commaOrRpar.treeNext?.treeParent?.appendNewlineMergingWhiteSpace(nextWhiteSpace, commaOrRpar.treeNext)
-                    } else {
-                        commaOrRpar.treeParent?.appendNewlineMergingWhiteSpace(nextWhiteSpace, commaOrRpar)
+
+        val classDefinitionNode = valueParameterList[0].treeParent?.treeParent
+        val colonNodes = classDefinitionNode?.getAllChildrenWithType(COLON)
+        val newlineBeforeSuperTypeList = colonNodes?.find {colonNode ->
+            val newlineNode = colonNode.treeNext
+            val superClassType = newlineNode?.treeNext
+            newlineNode?.text?.count { it == '\n' } == 1 && superClassType?.elementType == SUPER_TYPE_LIST
+        }
+
+        var areElementsCorrect = true
+        var valueParameter = valueParameterList[0].treeNext
+        while (valueParameter != null) {
+            val newlineNode = valueParameter.treeNext
+            val superClassType = valueParameter.treeNext?.treeNext
+            if (valueParameter.elementType != COMMA || newlineNode?.text?.count { it == '\n' } != 1 ||
+                    superClassType?.elementType != SUPER_TYPE_ENTRY) {
+                areElementsCorrect = false
+                break
+            }
+            valueParameter = superClassType?.treeNext
+        }
+        return newlineBeforeSuperTypeList != null && areElementsCorrect
+    }
+
+    @Suppress("TOO_LONG_FUNCTION", "ComplexMethod")
+    private fun handleValueParameterList(node: ASTNode, entryType: String) {
```

Comment:
Maybe extract smaller functions from this large function?

Reply:
What does value parameter list mean?

Reply:
Value parameter list - list of supertypes and interfaces, from which our class is inherited